The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8544317566/job/23410060480) [(plain)](https://github.com/rust-lang/rust/commit/420d1779375a529816991ddb705319991dcc5ab1/checks/23410060480/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#17 exporting to docker image format
#17 sending tarball 29.5s done
#17 DONE 34.6s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---

+ error: top-level or-patterns are not allowed in `let` bindings
+   --> $DIR/match-cfg-fake-edges.rs:35:9
+    |
+ LL |     let (y, _) | (_, y) = x;
+    |         ^^^^^^^^^^^^^^^ help: wrap the pattern in parentheses: `((y, _) | (_, y))`
+ 
+ warning: irrefutable `if let` pattern
+    |
+ LL |     if let _ = true {
+    |        ^^^^^^^^^^^^
+    |
---
1 error[E0382]: use of moved value: `x`
2   --> $DIR/match-cfg-fake-edges.rs:16:10
3    |

15 LL |         _ => drop(x.clone()),
17 
+ error[E0382]: use of moved value: `x`
+   --> $DIR/match-cfg-fake-edges.rs:24:10
+    |
+    |
+ LL |     let x = String::new();
+    |         - move occurs because `x` has type `String`, which does not implement the `Copy` trait
+ LL |         drop(x)
+    |              - value moved here
+ ...
+ LL |     drop(x);
---
18 error[E0382]: borrow of moved value: `x.0`
-   --> $DIR/match-cfg-fake-edges.rs:22:5
+   --> $DIR/match-cfg-fake-edges.rs:30:5
20    |
21 LL |         (y, _) | (_, y) => (),

31    |          +++
32 
33 error[E0382]: borrow of moved value: `x.1`
33 error[E0382]: borrow of moved value: `x.1`
-   --> $DIR/match-cfg-fake-edges.rs:24:5
+   --> $DIR/match-cfg-fake-edges.rs:32:5
35    |
36 LL |         (y, _) | (_, y) => (),


45 LL |         (y, _) | (_, ref y) => (),
47 
+ error[E0382]: borrow of moved value: `x.0`
+   --> $DIR/match-cfg-fake-edges.rs:36:5
+    |
+    |
+ LL |     let (y, _) | (_, y) = x;
+ LL |     &x.0;
+    |     ^^^^ value borrowed here after move
+    |
+    |
+    = note: move occurs because `x.0` has type `String`, which does not implement the `Copy` trait
+ help: borrow this binding in the pattern to avoid moving the value
+    |
+ LL |     let (ref y, _) | (_, y) = x;
+ 
+ error[E0382]: borrow of moved value: `x.1`
+   --> $DIR/match-cfg-fake-edges.rs:38:5
+    |
+    |
+ LL |     let (y, _) | (_, y) = x;
+ ...
+ LL |     &x.1;
+    |     ^^^^ value borrowed here after move
+    |
+    |
+    = note: move occurs because `x.1` has type `String`, which does not implement the `Copy` trait
+ help: borrow this binding in the pattern to avoid moving the value
+    |
+ LL |     let (y, _) | (_, ref y) = x;
+ 
48 error[E0381]: used binding `x` is possibly-uninitialized
-   --> $DIR/match-cfg-fake-edges.rs:58:19
+   --> $DIR/match-cfg-fake-edges.rs:72:19
---
-   --> $DIR/match-cfg-fake-edges.rs:85:22
+   --> $DIR/match-cfg-fake-edges.rs:99:22
94    |
95 LL |     let x = String::new();
96    |         - move occurs because `x` has type `String`, which does not implement the `Copy` trait
107    |                          ++++++++
108 
109 error[E0382]: use of moved value: `x`
-   --> $DIR/match-cfg-fake-edges.rs:100:22
-   --> $DIR/match-cfg-fake-edges.rs:100:22
+   --> $DIR/match-cfg-fake-edges.rs:114:22
111    |
112 LL |     let x = String::new();
113    |         - move occurs because `x` has type `String`, which does not implement the `Copy` trait

122 LL |         false if let Some(()) = { drop(x.clone()); Some(()) } => {},
124 
- error: aborting due to 8 previous errors
+ error: aborting due to 12 previous errors; 1 warning emitted
126 
---
--- stderr -------------------------------
error: top-level or-patterns are not allowed in `let` bindings
##[error]  --> /checkout/tests/ui/nll/match-cfg-fake-edges.rs:35:9
   |
LL |     let (y, _) | (_, y) = x;
   |         ^^^^^^^^^^^^^^^ help: wrap the pattern in parentheses: `((y, _) | (_, y))`

warning: irrefutable `if let` pattern
   |
LL |     if let _ = true {
   |        ^^^^^^^^^^^^
   |
---
error[E0382]: use of moved value: `x`
##[error]  --> /checkout/tests/ui/nll/match-cfg-fake-edges.rs:16:10
   |
LL |     let x = String::new();
   |         - move occurs because `x` has type `String`, which does not implement the `Copy` trait
LL |         _ => drop(x),
   |                   - value moved here
...
...
LL |     drop(x); //~ ERROR use of moved value
   |          ^ value used here after move
help: consider cloning the value if the performance cost is acceptable
   |
   |
LL |         _ => drop(x.clone()),

error[E0382]: use of moved value: `x`
##[error]  --> /checkout/tests/ui/nll/match-cfg-fake-edges.rs:24:10
   |
   |
LL |     let x = String::new();
   |         - move occurs because `x` has type `String`, which does not implement the `Copy` trait
Build completed unsuccessfully in 0:14:38
LL |         drop(x)
   |              - value moved here
...
...
LL |     drop(x); //~ ERROR use of moved value
   |          ^ value used here after move
help: consider cloning the value if the performance cost is acceptable
   |
LL |         drop(x.clone())
   |               ++++++++
   |               ++++++++

error[E0382]: borrow of moved value: `x.0`
##[error]  --> /checkout/tests/ui/nll/match-cfg-fake-edges.rs:30:5
   |
LL |         (y, _) | (_, y) => (),
LL |     }
LL |     }
LL |     &x.0; //~ ERROR borrow of moved value
   |
   |
   = note: move occurs because `x.0` has type `String`, which does not implement the `Copy` trait
help: borrow this binding in the pattern to avoid moving the value
   |
LL |         (ref y, _) | (_, y) => (),

error[E0382]: borrow of moved value: `x.1`
##[error]  --> /checkout/tests/ui/nll/match-cfg-fake-edges.rs:32:5
   |
   |
LL |         (y, _) | (_, y) => (),
...
...
LL |     &x.1; //~ ERROR borrow of moved value
   |
   |
   = note: move occurs because `x.1` has type `String`, which does not implement the `Copy` trait
help: borrow this binding in the pattern to avoid moving the value
   |
LL |         (y, _) | (_, ref y) => (),

error[E0382]: borrow of moved value: `x.0`
##[error]  --> /checkout/tests/ui/nll/match-cfg-fake-edges.rs:36:5
   |
   |
LL |     let (y, _) | (_, y) = x;
   |          - value moved here
LL |     &x.0; //~ ERROR borrow of moved value
   |
   |
   = note: move occurs because `x.0` has type `String`, which does not implement the `Copy` trait
help: borrow this binding in the pattern to avoid moving the value
   |
LL |     let (ref y, _) | (_, y) = x;

error[E0382]: borrow of moved value: `x.1`
##[error]  --> /checkout/tests/ui/nll/match-cfg-fake-edges.rs:38:5
   |
   |
LL |     let (y, _) | (_, y) = x;
...
...
LL |     &x.1; //~ ERROR borrow of moved value
   |
   |
   = note: move occurs because `x.1` has type `String`, which does not implement the `Copy` trait
help: borrow this binding in the pattern to avoid moving the value
   |
LL |     let (y, _) | (_, ref y) = x;

error[E0381]: used binding `x` is possibly-uninitialized
##[error]  --> /checkout/tests/ui/nll/match-cfg-fake-edges.rs:72:19
   |
   |
LL |     let x;
   |         - binding declared here but left uninitialized
...
LL |         _ => drop(x), //~ ERROR used binding `x` is possibly-uninitialized
   |         -         ^ `x` used here but it is possibly-uninitialized
   |         if this pattern is matched, `x` is not initialized

error[E0381]: used binding `x` isn't initialized
##[error]  --> /checkout/tests/ui/nll/match-cfg-fake-edges.rs:79:16
##[error]  --> /checkout/tests/ui/nll/match-cfg-fake-edges.rs:79:16
   |
LL |     let x;
   |         - binding declared here but left uninitialized
LL |     match y {
LL |         _ if { x = 2; true } => 1,
   |                ----- binding initialized here in some conditions
LL |         // Borrowck must not know the guard is always run.
LL |         _ if { x; false } => 2, //~ ERROR used binding `x` isn't initialized
   |                ^ `x` used here but it isn't initialized
help: consider assigning a value
   |
LL |     let x = 0;
   |           +++
   |           +++

error[E0381]: used binding `x` isn't initialized
##[error]  --> /checkout/tests/ui/nll/match-cfg-fake-edges.rs:86:31
   |
LL |     let x;
   |         - binding declared here but left uninitialized
LL |     match y {
LL |         _ if let Some(()) = { x = 2; Some(()) } => 1,
   |                               ----- binding initialized here in some conditions
LL |         _ if let Some(()) = { x; None } => 2, //~ ERROR used binding `x` isn't initialized
   |                               ^ `x` used here but it isn't initialized
help: consider assigning a value
   |
LL |     let x = 0;
   |           +++
   |           +++

error[E0382]: use of moved value: `x`
##[error]  --> /checkout/tests/ui/nll/match-cfg-fake-edges.rs:99:22
   |
LL |     let x = String::new();
   |         - move occurs because `x` has type `String`, which does not implement the `Copy` trait
LL |     match y {
LL |         false if { drop(x); true } => {},
   |                         - value moved here
LL |         // Borrowck must not know the guard is not run in the `true` case.
LL |         true => drop(x), //~ ERROR use of moved value: `x`
   |                      ^ value used here after move
help: consider cloning the value if the performance cost is acceptable
   |
   |
LL |         false if { drop(x.clone()); true } => {},

error[E0382]: use of moved value: `x`
##[error]  --> /checkout/tests/ui/nll/match-cfg-fake-edges.rs:114:22
   |
   |
LL |     let x = String::new();
   |         - move occurs because `x` has type `String`, which does not implement the `Copy` trait
LL |     match y {
LL |         false if let Some(()) = { drop(x); Some(()) } => {},
   |                                        - value moved here
LL |         true => drop(x), //~ ERROR use of moved value: `x`
   |                      ^ value used here after move
help: consider cloning the value if the performance cost is acceptable
   |
   |
LL |         false if let Some(()) = { drop(x.clone()); Some(()) } => {},

error: aborting due to 12 previous errors; 1 warning emitted

Some errors have detailed explanations: E0381, E0382.
```

</details>

--- a/tests/ui/nll/match-cfg-fake-edges.rs
+++ b/tests/ui/nll/match-cfg-fake-edges.rs
@@ -8,17 +8,31 @@ fn all_patterns_are_tested() {
     // Even though `x` is never actually moved out of, we don't want borrowck results to be based on
     // whether MIR lowering reveals which patterns are unreachable.
     let x = String::new();
-    let _ = match true {
+    match true {
         _ => {},
         _ => drop(x),
-    };
+    }
     // Borrowck must not know the second arm is never run.
+    drop(x); //~ ERROR use of moved value
+
+    let x = String::new();
+    if let _ = true { //~ WARN irrefutable
+    } else {
+        drop(x)
+    }
+    // Borrowck must not know the else branch is never run.
     drop(x); //~ ERROR use of moved value
 
     let x = (String::new(), String::new());
     match x {
         (y, _) | (_, y) => (),
     }
+    &x.0; //~ ERROR borrow of moved value
+    // Borrowck must not know the second pattern never matches.
+    &x.1; //~ ERROR borrow of moved value
+
+    let x = (String::new(), String::new());
+    let ((y, _) | (_, y)) = x;
     &x.0; //~ ERROR borrow of moved value
     // Borrowck must not know the second pattern never matches.
     &x.1; //~ ERROR borrow of moved value

--- a/tests/ui/nll/match-cfg-fake-edges.stderr
+++ b/tests/ui/nll/match-cfg-fake-edges.stderr
@@ -1,3 +1,13 @@
+warning: irrefutable `if let` pattern
+  --> $DIR/match-cfg-fake-edges.rs:19:8
+   |
+LL |     if let _ = true {
+   |        ^^^^^^^^^^^^
+   |
+   = note: this pattern will always match, so the `if let` is useless
+   = help: consider replacing the `if let` with a `let`
+   = note: `#[warn(irrefutable_let_patterns)]` on by default
+
 error[E0382]: use of moved value: `x`
   --> $DIR/match-cfg-fake-edges.rs:16:10
    |
@@ -15,8 +25,25 @@ help: consider cloning the value if the performance cost is acceptable
 LL |         _ => drop(x.clone()),
    |                    ++++++++
 
+error[E0382]: use of moved value: `x`
+  --> $DIR/match-cfg-fake-edges.rs:24:10
+   |
+LL |     let x = String::new();
+   |         - move occurs because `x` has type `String`, which does not implement the `Copy` trait
+...
+LL |         drop(x)
+   |              - value moved here
+...
+LL |     drop(x);
+   |          ^ value used here after move
+   |
+help: consider cloning the value if the performance cost is acceptable
+   |
+LL |         drop(x.clone())
+   |               ++++++++
+
 error[E0382]: borrow of moved value: `x.0`
-  --> $DIR/match-cfg-fake-edges.rs:22:5
+  --> $DIR/match-cfg-fake-edges.rs:30:5
    |
 LL |         (y, _) | (_, y) => (),
    |          - value moved here
@@ -31,7 +58,7 @@ LL |         (ref y, _) | (_, y) => (),
    |          +++
 
 error[E0382]: borrow of moved value: `x.1`
-  --> $DIR/match-cfg-fake-edges.rs:24:5
+  --> $DIR/match-cfg-fake-edges.rs:32:5
    |
 LL |         (y, _) | (_, y) => (),
    |                      - value moved here
@@ -45,8 +72,37 @@ help: borrow this binding in the pattern to avoid moving the value
 LL |         (y, _) | (_, ref y) => (),
    |                      +++
 
+error[E0382]: borrow of moved value: `x.0`
+  --> $DIR/match-cfg-fake-edges.rs:36:5
+   |
+LL |     let ((y, _) | (_, y)) = x;
+   |           - value moved here
+LL |     &x.0;
+   |     ^^^^ value borrowed here after move
+   |
+   = note: move occurs because `x.0` has type `String`, which does not implement the `Copy` trait
+help: borrow this binding in the pattern to avoid moving the value
+   |
+LL |     let ((ref y, _) | (_, y)) = x;
+   |           +++
+
+error[E0382]: borrow of moved value: `x.1`
+  --> $DIR/match-cfg-fake-edges.rs:38:5
+   |
+LL |     let ((y, _) | (_, y)) = x;
+   |                       - value moved here
+...
+LL |     &x.1;
+   |     ^^^^ value borrowed here after move
+   |
+   = note: move occurs because `x.1` has type `String`, which does not implement the `Copy` trait
+help: borrow this binding in the pattern to avoid moving the value
+   |
+LL |     let ((y, _) | (_, ref y)) = x;
+   |                       +++
+
 error[E0381]: used binding `x` is possibly-uninitialized
-  --> $DIR/match-cfg-fake-edges.rs:58:19
+  --> $DIR/match-cfg-fake-edges.rs:72:19
    |
 LL |     let x;
    |         - binding declared here but left uninitialized
@@ -57,7 +113,7 @@ LL |         _ => drop(x),
    |         if this pattern is matched, `x` is not initialized
 
 error[E0381]: used binding `x` isn't initialized
-  --> $DIR/match-cfg-fake-edges.rs:65:16
+  --> $DIR/match-cfg-fake-edges.rs:79:16
    |
 LL |     let x;
    |         - binding declared here but left uninitialized
@@ -74,7 +130,7 @@ LL |     let x = 0;
    |           +++
 
 error[E0381]: used binding `x` isn't initialized
-  --> $DIR/match-cfg-fake-edges.rs:72:31
+  --> $DIR/match-cfg-fake-edges.rs:86:31
    |
 LL |     let x;
    |         - binding declared here but left uninitialized
@@ -90,7 +146,7 @@ LL |     let x = 0;
    |           +++
 
 error[E0382]: use of moved value: `x`
-  --> $DIR/match-cfg-fake-edges.rs:85:22
+  --> $DIR/match-cfg-fake-edges.rs:99:22
    |
 LL |     let x = String::new();
    |         - move occurs because `x` has type `String`, which does not implement the `Copy` trait
@@ -107,7 +163,7 @@ LL |         false if { drop(x.clone()); true } => {},
    |                          ++++++++
 
 error[E0382]: use of moved value: `x`
-  --> $DIR/match-cfg-fake-edges.rs:100:22
+  --> $DIR/match-cfg-fake-edges.rs:114:22
    |
 LL |     let x = String::new();
    |         - move occurs because `x` has type `String`, which does not implement the `Copy` trait
@@ -122,7 +178,7 @@ help: consider cloning the value if the performance cost is acceptable
 LL |         false if let Some(()) = { drop(x.clone()); Some(()) } => {},
    |                                         ++++++++
 
-error: aborting due to 8 previous errors
+error: aborting due to 11 previous errors; 1 warning emitted
 
 Some errors have detailed explanations: E0381, E0382.
 For more information about an error, try `rustc --explain E0381`.

